File: sdk/typescript/src/adapters/crewai.ts
/**
 * CrewAI adapter for ADP v0.3.0.
 *
 * Converts between ADP manifests and CrewAI configurations.
 */

import { AdapterBase } from "./base";
import { AdapterRegistry } from "./registry";
import { ADP } from "../adp";

export class CrewAIAdapter extends AdapterBase {
  readonly frameworkId = "crewai";

  constructor() {
    super();
    AdapterRegistry.register(CrewAIAdapter);
  }

  export(manifest: ADP): Record<string, unknown> {
    const data = manifest as Record<string, unknown>;

    const flow = (data.flow as Record<string, unknown>) ?? {};
    const graph = (flow.graph as Record<string, unknown>) ?? {};
    const nodes = (graph.nodes as Array<Record<string, unknown>>) ?? [];

    // Map ADP nodes to CrewAI agents and tasks
    const agents: Record<string, Record<string, unknown>> = {};
    const tasks: Array<Record<string, unknown>> = [];

    for (const node of nodes) {
      const nid = (node.id as string) ?? "";
      const kind = (node.kind as string) ?? "";
      const role = (node.label as string) ?? nid;

      if (kind === "llm") {
        agents[nid] = {
          role: role,
          llm: node.model_ref as string ?? "gpt-4",
        };
      } else if (kind === "tool") {
        // Tools are mapped to agent tools
        // In CrewAI, tools are defined at the agent level
      }
    }

    // Build crew configuration
    const crew: Record<string, unknown> = {
      agents: Object.values(agents),
      tasks: tasks,
      process: "sequential", // default
    };

    // Apply adapter hints
    const adapterHints = (
      (data.runtime as Record<string, unknown>)?.adapter_hints as Record<string, unknown>
    )?.crewai as Record<string, unknown> | undefined;

    if (adapterHints) {
      if (adapterHints.process !== undefined) {
        crew.process = adapterHints.process;
      }
      if (adapterHints.max_rpm !== undefined) {
        crew.max_rpm = adapterHints.max_rpm;
      }
    }

    return crew;
  }

  importFrom(config: Record<string, unknown>): ADP {
    const agents = (config.agents as Array<Record<string, unknown>>) ?? [];
    const tasks = (config.tasks as Array<Record<string, unknown>>) ?? [];
    const process = (config.process as string) ?? "sequential";

    const flowNodes: Array<Record<string, unknown>> = [];
    const flowEdges: Array<Record<string, unknown>> = [];

    // Convert agents to nodes
    for (const agent of agents) {
      const role = (agent.role as string) ?? "";
      const llm = (agent.llm as string) ?? "";

      flowNodes.push({
        id: role.toLowerCase().replace(/\s+/g, "_"),
        kind: "llm",
        label: role,
        model_ref: llm,
      });
    }

    // Connect nodes sequentially
    for (let i = 0; i < flowNodes.length - 1; i++) {
      flowEdges.push({
        from: flowNodes[i].id,
        to: flowNodes[i + 1].id,
      });
    }

    // Extract adapter hints
    const adapterHints: Record<string, unknown> = {};
    if (process) {
      adapterHints.process = process;
    }
    if (config.max_rpm !== undefined) {
      adapterHints.max_rpm = config.max_rpm;
    }

    const startNodes = flowNodes.length > 0 ? [flowNodes[0].id as string] : [];
    const endNodes = flowNodes.length > 0 ? [flowNodes[flowNodes.length - 1].id as string] : [];

    return {
      adp_version: "0.3.0",
      id: (config.id as string) ?? "imported-from-crewai",
      runtime: {
        execution: [
          {
            backend: "python",
            id: "crewai",
            module: "crewai",
          },
        ],
      },
      flow: {
        id: "imported-flow",
        graph: {
          nodes: flowNodes,
          edges: flowEdges,
          start_nodes: startNodes,
          end_nodes: endNodes,
        },
        extensions: adapterHints && Object.keys(adapterHints).length > 0 ? { crewai: adapterHints } : undefined,
      },
      evaluation: { suites: [] },
      extensions: {
        source_framework: "crewai",
      },
    } as unknown as ADP;
  }

  roundtripFidelity(): Record<string, string> {
    return {
      ...super.roundtripFidelity(),
      "tools.policy": "faithful", // via rate_limit -> max_rpm
    };
  }
}

File: sdk/python/adp_sdk/adapters/crewai.py
"""CrewAI adapter for ADP v0.3.0.

Converts between ADP manifests and CrewAI configurations.
"""

from typing import Any

from .base import AdapterBase
from .registry import AdapterRegistry
from ..adp_model import ADP


class CrewAIAdapter(AdapterBase):
    """Adapter for CrewAI framework.

    Converts ADP agents and tasks to CrewAI configurations.
    """

    framework_id = "crewai"

    def __init__(self) -> None:
        AdapterRegistry.register(self.__class__)

    def export(self, manifest: ADP) -> dict[str, Any]:
        """Export ADP manifest to CrewAI config."""
        data = manifest.model_dump(by_alias=True, exclude_none=True)

        flow = data.get("flow", {})
        graph = flow.get("graph", {})
        nodes = graph.get("nodes", [])

        # Map ADP nodes to CrewAI agents and tasks
        agents: dict[str, dict[str, Any]] = {}
        tasks: list[dict[str, Any]] = []

        for node in nodes:
            nid = node.get("id", "")
            kind = node.get("kind", "")
            role = node.get("label", nid)

            if kind == "llm":
                agents[nid] = {
                    "role": role,
                    "llm": node.get("model_ref", "gpt-4"),
                }
            elif kind == "tool":
                # Tools are mapped to agent tools
                pass

        # Build crew configuration
        crew: dict[str, Any] = {
            "agents": list(agents.values()),
            "tasks": tasks,
            "process": "sequential",  # default
        }

        # Apply adapter hints
        adapter_hints = (data.get("runtime", {}).get("adapter_hints", {})
                         .get("crewai", {}))

        if process := adapter_hints.get("process"):
            crew["process"] = process
        if max_rpm := adapter_hints.get("max_rpm"):
            crew["max_rpm"] = max_rpm

        return crew

    def import_from(self, config: dict[str, Any]) -> ADP:
        """Import CrewAI config to ADP manifest."""
        agents = config.get("agents", [])
        process = config.get("process", "sequential")

        flow_nodes = []
        flow_edges = []

        # Convert agents to nodes
        for agent in agents:
            role = agent.get("role", "")
            llm = agent.get("llm", "")

            flow_nodes.append({
                "id": role.lower().replace(" ", "_"),
                "kind": "llm",
                "label": role,
                "model_ref": llm,
            })

        # Connect nodes sequentially
        for i in range(len(flow_nodes) - 1):
            flow_edges.append({
                "from": flow_nodes[i]["id"],
                "to": flow_nodes[i + 1]["id"],
            })

        # Extract adapter hints
        adapter_hints = {}
        if process:
            adapter_hints["process"] = process
        if max_rpm := config.get("max_rpm"):
            adapter_hints["max_rpm"] = max_rpm

        from ..adp_model import (
            RuntimeEntry,
            RuntimeModel,
            GraphModel,
            FlowModel,
        )

        return ADP(
            adp_version="0.3.0",
            id=config.get("id", "imported-from-crewai"),
            runtime=RuntimeModel(execution=[RuntimeEntry(
                backend="python",
                id="crewai",
                module="crewai",
            )]),
            flow=FlowModel(
                id="imported-flow",
                graph=GraphModel(
                    nodes=flow_nodes,
                    edges=flow_edges,
                    start_nodes=[flow_nodes[0]["id"]] if flow_nodes else [],
                    end_nodes=[flow_nodes[-1]["id"]] if flow_nodes else [],
                ),
                extensions={"crewai": adapter_hints} if adapter_hints else None,
            ),
            extensions={"source_framework": "crewai"},
        )

    def roundtrip_fidelity(self) -> dict[str, str]:
        return {
            **super().roundtrip_fidelity(),
            "tools.policy": "faithful",  # via rate_limit -> max_rpm
        }
